File: packages/flags-react-native/src/user-context.ts
/**
 * SignaKitUserContext — represents a user for flag evaluation in a
 * React Native runtime. Auto-fires `$exposure` events for non-targeted
 * decisions, deduplicated per app session via an in-memory Map.
 */

import {
  MAX_EVENT_KEY_LENGTH,
  MAX_USER_ID_LENGTH,
  MAX_METADATA_SIZE_BYTES,
  MAX_ATTRIBUTES_COUNT,
  MAX_ATTRIBUTE_KEY_LENGTH,
  MAX_ATTRIBUTE_VALUE_LENGTH,
} from './constants'
import type {
  SignaKitDecision,
  SignaKitDecisions,
  SignaKitEvent,
  TrackEventOptions,
  UserAttributes,
} from './types'
import type { SignaKitClient } from './client'

/**
 * Sanitize attributes to enforce size limits.
 */
export function sanitizeAttributes(
  attributes: UserAttributes | undefined
): UserAttributes | undefined {
  if (!attributes || Object.keys(attributes).length === 0) {
    return undefined
  }

  const sanitized: UserAttributes = {}
  const keys = Object.keys(attributes).slice(0, MAX_ATTRIBUTES_COUNT)

  for (const key of keys) {
    const sanitizedKey = key.slice(0, MAX_ATTRIBUTE_KEY_LENGTH)
    const value = attributes[key]

    if (value === undefined) {
      continue
    } else if (typeof value === 'string') {
      sanitized[sanitizedKey] = value.slice(0, MAX_ATTRIBUTE_VALUE_LENGTH)
    } else if (Array.isArray(value)) {
      sanitized[sanitizedKey] = (value as string[])
        .slice(0, 100)
        .map((v) => v.slice(0, MAX_ATTRIBUTE_VALUE_LENGTH))
    } else {
      sanitized[sanitizedKey] = value
    }
  }

  return sanitized
}

export class SignaKitUserContext {
  private client: SignaKitClient
  private cachedDecisions: Record<string, string> | null = null
  readonly userId: string
  readonly attributes: UserAttributes

  constructor(client: SignaKitClient, userId: string, attributes: UserAttributes = {}) {
    this.client = client
    this.userId = userId
    // Strip $userAgent — kept for parity with browser SDK but not used for
    // any runtime detection in React Native.
    const { $userAgent: _ignored, ...rest } = attributes
    void _ignored
    this.attributes = rest
  }

  /**
   * Fire-and-forget exposure tracking for non-targeted decisions.
   * Deduplicated per app session via the client's in-memory Map.
   */
  private _trackExposure(decision: SignaKitDecision): void {
    // Targeted rules are simple feature-flag rollouts — no experiment to
    // attribute, so skip exposure entirely.
    if (decision.ruleType === 'targeted') return

    const dedupKey = `${decision.flagKey}:${this.userId}`
    if (this.client._hasExposed(dedupKey)) return

    const event: SignaKitEvent = {
      eventKey: '$exposure',
      userId: this.userId.slice(0, MAX_USER_ID_LENGTH),
      timestamp: new Date().toISOString(),
      decisions: { [decision.flagKey]: decision.variationKey },
      metadata: {
        flagKey: decision.flagKey,
        variationKey: decision.variationKey,
        ruleKey: decision.ruleKey,
      },
    }

    const sanitizedAttrs = sanitizeAttributes(this.attributes)
    if (sanitizedAttrs) {
      event.attributes = sanitizedAttrs
    }

    // Mark as sent before firing (prevent re-entry).
    this.client._markExposed(dedupKey)

    this.client._sendEvent(event).catch(() => {
      // Silently ignore - exposure tracking should not break the app.
    })
  }

  /**
   * Evaluate a single flag for this user.
   */
  decide(flagKey: string): SignaKitDecision | null {
    const decision = this.client._evaluateFlag(flagKey, this.userId, this.attributes)

    if (decision) {
      if (!this.cachedDecisions) {
        this.cachedDecisions = {}
      }
      this.cachedDecisions[flagKey] = decision.variationKey
      this._trackExposure(decision)
    }

    return decision
  }

  /**
   * Evaluate all flags for this user.
   */
  decideAll(): SignaKitDecisions {
    const decisions = this.client._evaluateAllFlags(this.userId, this.attributes)

    this.cachedDecisions = {}
    for (const [flagKey, decision] of Object.entries(decisions)) {
      this.cachedDecisions[flagKey] = decision.variationKey
      this._trackExposure(decision)
    }

    return decisions
  }

  /**
   * Track a conversion event for this user.
   */
  async trackEvent(eventKey: string, options: TrackEventOptions = {}): Promise<void> {
    const sanitizedEventKey = eventKey.slice(0, MAX_EVENT_KEY_LENGTH)
    const sanitizedUserId = this.userId.slice(0, MAX_USER_ID_LENGTH)

    const event: SignaKitEvent = {
      eventKey: sanitizedEventKey,
      userId: sanitizedUserId,
      timestamp: new Date().toISOString(),
    }

    const sanitizedAttrs = sanitizeAttributes(this.attributes)
    if (sanitizedAttrs) {
      event.attributes = sanitizedAttrs
    }

    if (this.cachedDecisions && Object.keys(this.cachedDecisions).length > 0) {
      event.decisions = this.cachedDecisions
    }

    if (options.value !== undefined) {
      event.value = options.value
    }

    if (options.metadata) {
      const metadataStr = JSON.stringify(options.metadata)
      if (metadataStr.length <= MAX_METADATA_SIZE_BYTES) {
        event.metadata = options.metadata
      } else {
        console.warn(
          `[SignaKit] metadata exceeds ${MAX_METADATA_SIZE_BYTES} bytes (${metadataStr.length}), dropping`
        )
      }
    }

    await this.client._sendEvent(event)
  }
}

File: packages/flags-python/src/signakit_flags/user_context.py
"""User context — created via :class:`SignaKitClient.create_user_context`."""

from __future__ import annotations

import asyncio
import json
import logging
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any

from .bot_patterns import is_bot
from .constants import (
    MAX_ATTRIBUTE_KEY_LENGTH,
    MAX_ATTRIBUTE_VALUE_LENGTH,
    MAX_ATTRIBUTES_COUNT,
    MAX_EVENT_KEY_LENGTH,
    MAX_METADATA_SIZE_BYTES,
    MAX_USER_ID_LENGTH,
)
from .types import (
    Decision,
    Decisions,
    TrackEventOptions,
    UserAttributes,
)

if TYPE_CHECKING:
    from .client import SignaKitClient

_logger = logging.getLogger(__name__)


def _sanitize_attributes(
    attributes: UserAttributes | None,
) -> UserAttributes | None:
    """Truncate keys/values and cap attribute count, mirroring the JS SDK."""
    if not attributes:
        return None

    sanitized: UserAttributes = {}
    for key in list(attributes.keys())[:MAX_ATTRIBUTES_COUNT]:
        value = attributes[key]
        if value is None:  # pragma: no cover — defensive
            continue
        truncated_key = key[:MAX_ATTRIBUTE_KEY_LENGTH]
        if isinstance(value, str):
            sanitized[truncated_key] = value[:MAX_ATTRIBUTE_VALUE_LENGTH]
        elif isinstance(value, list):
            sanitized[truncated_key] = [
                v[:MAX_ATTRIBUTE_VALUE_LENGTH] for v in value[:100]
            ]
        else:
            sanitized[truncated_key] = value
    return sanitized or None


def _now_iso() -> str:
    """ISO 8601 timestamp with millisecond precision and ``Z`` suffix."""
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


class SignaKitUserContext:
    """Represents a user for flag evaluation and event tracking."""

    def __init__(
        self,
        client: SignaKitClient,
        user_id: str,
        attributes: UserAttributes | None = None,
    ) -> None:
        self._client = client
        self.user_id: str = user_id

        attributes = dict(attributes) if attributes else {}
        user_agent_raw = attributes.pop("$userAgent", None)
        user_agent = user_agent_raw if isinstance(user_agent_raw, str) else None
        self._is_bot: bool = is_bot(user_agent)
        self.attributes: UserAttributes = attributes
        self._cached_decisions: dict[str, str] = {}

    # ---- exposure helpers --------------------------------------------------

    def _build_exposure_event(self, decision: Decision) -> dict[str, Any]:
        event: dict[str, Any] = {
            "eventKey": "$exposure",
            "userId": self.user_id[:MAX_USER_ID_LENGTH],
            "timestamp": _now_iso(),
            "decisions": {decision.flag_key: decision.variation_key},
            "metadata": {
                "flagKey": decision.flag_key,
                "variationKey": decision.variation_key,
                "ruleKey": decision.rule_key,
            },
        }
        sanitized = _sanitize_attributes(self.attributes)
        if sanitized:
            event["attributes"] = sanitized
        return event

    def _send_exposure(self, decision: Decision) -> None:
        """Fire-and-forget exposure event.

        Skipped for ``targeted`` rules — those are simple feature-flag rollouts
        with no experiment to attribute, so exposures would just be noise.
        """
        if decision.rule_type == "targeted":
            return

        event = self._build_exposure_event(decision)

        # Prefer scheduling on a running event loop; fall back to synchronous
        # transport so this works in plain scripts too.
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            try:
                self._client._send_event_sync(event)
            except Exception:  # noqa: BLE001 — fire-and-forget
                _logger.debug("[SignaKit] exposure send failed", exc_info=True)
            return

        async def _runner() -> None:
            try:
                await self._client._send_event(event)
            except Exception:  # noqa: BLE001 — fire-and-forget
                _logger.debug("[SignaKit] exposure send failed", exc_info=True)

        loop.create_task(_runner())

    # ---- public API --------------------------------------------------------

    def decide(self, flag_key: str) -> Decision | None:
        """Evaluate a single flag for this user."""
        if self._is_bot:
            return Decision(
                flag_key=flag_key,
                variation_key="off",
                enabled=False,
                rule_key=None,
                rule_type=None,
                variables={},
            )

        decision = self._client._evaluate_flag(flag_key, self.user_id, self.attributes)
        if decision is not None:
            self._cached_decisions[flag_key] = decision.variation_key
            self._send_exposure(decision)
        return decision

    def decide_all(self) -> Decisions:
        """Evaluate every non-archived flag for this user."""
        if self._is_bot:
            return self._client._get_bot_decisions()

        decisions = self._client._evaluate_all_flags(self.user_id, self.attributes)
        self._cached_decisions = {}
        for flag_key, decision in decisions.items():
            self._cached_decisions[flag_key] = decision.variation_key
            self._send_exposure(decision)
        return decisions

    async def track_event(
        self,
        event_key: str,
        value: float | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        """Track a conversion event for this user (sent immediately)."""
        if self._is_bot:
            return

        event: dict[str, Any] = {
            "eventKey": event_key[:MAX_EVENT_KEY_LENGTH],
            "userId": self.user_id[:MAX_USER_ID_LENGTH],
            "timestamp": _now_iso(),
        }

        sanitized = _sanitize_attributes(self.attributes)
        if sanitized:
            event["attributes"] = sanitized

        if self._cached_decisions:
            event["decisions"] = dict(self._cached_decisions)

        opts = TrackEventOptions(value=value, metadata=metadata)
        if opts.value is not None:
            event["value"] = opts.value

        if opts.metadata is not None:
            metadata_str = json.dumps(opts.metadata)
            if len(metadata_str) <= MAX_METADATA_SIZE_BYTES:
                event["metadata"] = opts.metadata
            else:
                _logger.warning(
                    "[SignaKit] metadata exceeds %d bytes (%d), dropping",
                    MAX_METADATA_SIZE_BYTES,
                    len(metadata_str),
                )

        await self._client._send_event(event)
